examples/ts/strict-count/src/models/util.d.ts

```diff
@@ -0,0 +1,47 @@
+type ModelReducers<S = any> = {
+  [key: string]: (state: S, payload: any) => S;
+};
+type ModelEffects<RootState = any> = {
+  [key: string]: (payload: any, rootState: RootState) => Promise<void>;
+};
+export type ModelConfig<S = any> = {
+  state: S;
+  reducers: ModelReducers<S>;
+  effects: (dispatch: any) => ModelEffects<any>;
+};
+export type Models<M=any> = {
+  [key: string]: ModelConfig<M>;
+};
+export type RematchRootState<M extends Models> = {
+  [modelKey in keyof M]: M[modelKey]["state"]
+};
+
+export type ReducerFromModel<R extends ModelReducers> = {
+  [reducerKey in keyof R]: Reducer2connect<R[reducerKey]>
+};
+export type EffectFromModel<E extends ModelEffects> = {
+  [effectKey in keyof E]: Effect2connect<E[effectKey]>
+};
+
+type Reducer2connect<R extends Function> = R extends (
+  state: infer S,
+  ...payload: infer P
+) => any ? (...payload: P) => S: () => void
+
+
+type Effect2connect<E extends Function> = E extends (
+  payload: infer P,
+  ...args: any
+) => any
+  ? (payload: P) => Promise<void>
```

Comment:
`  ? (payload: P) => Promise<void>`
should be
`  ? (payload?: P) => Promise<void>`

Because sometimes an effect may only use the second argument `rootState`:

```
effects: (dispatch: RootDispatch) => ({
    async foo(_payload: any, rootState: RootState) {
        doSomethingWith(rootState);
        dispatch.dolphins.setSomething();
    }
});
```

In this case when we call `foo` in an react component, we'll not pass any arguments:

```ts
foo();
```

So the payload argument should be optional

Reply:
I fix some problems, I think it's ok now

Reply:
This issue still exist, I created a demo in https://github.com/xcatliu/rematch/tree/hardfist/yj-dev/examples/ts/strict-count
You can clone it and repro this issue

![image](https://user-images.githubusercontent.com/5453359/64397682-a85e4b80-d094-11e9-9b4d-9135c89ac77a.png)

Reply:
Add `?` in line 45 will solve this issue

Reply:
oh I know what you mean now, I don't think   `doubleAsync()` should pass the type check, you can just call `doubleAsync(undefined) `, consider the following two methods
```
async doubleAsync1(_payload: any, rootState: RootState) {
			await delay(500)
			dispatch.sharks.increment(rootState.sharks)
}
```
```
async doubleAsync2(_payload: any, rootState: RootState) {
                        console.log('payload:', payload);
			await delay(500)
			dispatch.sharks.increment(rootState.sharks)
}
```
call `doubleAsync1()` ignoring first parameters may seems ok ,but it's not ok to call `doubleAsync2()`, so allow call `doubelAsync()` may cause problem for doubleAsync2